### DISCUSSION TITLE
Improvements to accounts-passwordless

### DISCUSSION THREAD

**Discussion by msavin** [meteor/meteor#12270] [Feature Requests]
I've just begun using accounts-passwordless - very excited for the application of it, but also very surprised at how the package is implemented. It feels like a drop from the quality that Meteor typically provides.

Here's my top three issues:

### 1 - Arbitrary requirement for email

For some reason, the package requires you to pass in an email for account creation if the account does not exist. However, it's not the same email as the selector. It seems like something can go wrong here from a security standpoint - also - just sending duplicate data feels weird. 

I understand that the selector might not always be an email, but if it is, we should rely on it and not require email to be passed in.

```Accounts.requestLoginTokenForUser({
	selector: data.email,
	userData: {
		email: data.email
	}
}, function (e,r) {
	// ... 
})
```

### 2 - Inform client if the sign in was a new account or old account

Currently, it doesn't tell you the nature of the passwordless authentication - particularly, if its a new sign up or an old sign in. It can pass a boolean value into the callback, if true, its a new account, otherwise, it's a sign in.

### 3- Ability to generate login-in tokens manually

For most my apps, we have custom email systems with templates, delivery, etc, automated. It would be much nicer to just call `Accounts.generateSignInToken(userId)` and handle the logic yourself for special situations. 

### 4 - Ability to support multiple login codes

Currently, the package is using $set to set the latest sign-in code on the user document. However, let's say a user fires off a sign-in event on their desktop and mobile device at the same time... only one of the codes they receive would work since the old one would be cleared by $set. 

Proposed solution: 
- create a collection for every sign in request, and clear documents from it when the sign-in tokens expire. Alternatively, an admin may want to keep these documents for security purposes.
- each document in this collection should have a `requestId` and a `signInCode`
- `Account.requestLoginTokenForUser` should generate the requestId, and `Account.passwordlessLoginWithToken` should verify the sign-in code entered by the user against the `requestId`. If the two match, they should be signed in. 

### Edit: 5 - Rate Limiting Requests

Are there any controls to request how many tokens were requested and to limit it against spam protection (i.e. you can request 15 tokens per hour)

**Comment by paulincai** [comment]
I support your perception.
Here: https://forums.meteor.com/t/a-call-to-meteor-team-to-improve-quality/58416
Here: https://github.com/meteor/meteor/issues/12077
and here: https://github.com/meteor/meteor/issues/12074

Meanwhile I could see a lot of improvements and we are already 1 version far from July when I faced this issues.

I ended up forking it and unfortunately I am due to push back some PRs.


You point 1. 
I agree, I guess there are not many situations you'd want to use that and perhaps it only makes sense if your selector is not an email. I, by default send the same email on both parameters from the client form but, you are right, need to relook into this.

Point number 2.

```
 Accounts.requestLoginTokenForUser({
      selector: { email },
      userData: { email },
      options: { extra }
    }, (error, { isNewUser } = {}) => {}) // isNewUser tells you whether this is a new account. I send them to an on-boarding workflow if new or just send them to the profile if it's an old user.
```
Point 3:
I plugged this generator  into my fork and now I generate my 4 digit numerical codes. I actually requested for a way to add your own generator via options.

```
function generateSequence () {
  return Math.floor(1000 + Math.random() * 9000)?.toString()
}
```

Point 4. My opinion is that it adds some complexity which is not necessary. By concept, passwordless tokens are supposed to have a short life. The shorter, the better. I'd even go down to 2 min and not more than 5. You have multiple devices, you can authenticate them 1 by  1 with no issues. You can even authenticate a mobile phone and authenticate everything else with QR codes like Whatsapp does and most crypto apps. I feel that token needs to die as soon as possible.

**Comment by msavin** [comment]
Would love to see that PR!

On Point 3 

Agree on that - but also - I just mean't like there should be a server-side function for the developer to generate a useable token and to then share that with the end user through their own means. I suspect most developers do not use Meteor's email system.

On Point 4

There's other situations where I've observed problems with the "keep only 1 to token alive" approach. For example: 
 - you enter your email into the app
 - you go to your inbox to get the token
 - you forget where the app browser tab is
 - you open open a new browser tab with the same app 
 - you enter your email again
 - you try the pin that you had just received rather than wait for a new one

Minor... but I can see a good number of instances where not keeping track of login tokens can lead to bad UX... and it could and should work _perfectly_.

I agree on the need to expire tokens as soon as possible - I like that there is a configurable field. Oddly, the field is set to hour... feels like minutes is much more reasonable. If you want 10 minutes, you have to set like 1.666666667...

**Comment by msavin** [comment]
Another major problem: there's no way to change the users emails 

Ideally, we should be able to add and remove emails from the account through the email token verification... could be a third party package of course but feels 'required' for a core package like this

**Comment by paulincai** [comment]
I take it you are managing business apps where you provide passwords to your users. To be frank, I thought everyone is using Meteor's email system or at least all those that use the Accounts system. I have just finished a high output email system based on Meteor and now emails of multiple platforms I work with go out via this email sender server with AWS Simple Email Service for SMTP. I can now set the volume as per the SMTP provider and send in batched. AWS SES for instance has 14 emails per second by default and this can be increased upon request. I could probably set up another SMTP and send 2-3 batches per second via 2-3 SMTPs if necessary.  But this is a different conversation.

I thing the UX concepts have changed lately. For instance in some browsers if i have domain.com open in a browser and I open a new tap and type in domain.com it asks me if i want to open it in the tab where it already exists.
On mobile, you never have to leave your App/browser to get the code. I attach for you a notification as received on my Android phone. I just pull down to get the notifications screen expand the notification and get the code. Pull up to close the notifications overlay of Android. This is similar to what would have happened if I sent it by SMS.

I only use one email per user. I allow the use of extra emails for correspondence for instance to receive updates. Log in with my personal email cause I will always have it but send news/updates to my business email ... which ... I might change my job one day. I personally do not allow users to change their own login emails. I keep this as a Support function.

I think passwordless is great and I am planning to use it for other forms of authentication such as non-financial transactions.

![af239d0a-fe9b-4776-ba83-186646a00656](https://user-images.githubusercontent.com/26188915/198875662-76b97327-4c0d-4a54-9bea-76910ef927ba.jpg)

I understand what would be necessary for now is, and please add to the list:

1. Use own token generation function or build 2-3 functions so that I can set up what type of token I want.
2. Expiry in minutes rather than hours.
